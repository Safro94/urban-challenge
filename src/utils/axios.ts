import axios, { AxiosError, AxiosRequestConfig } from 'axios';
import https from 'https';

const opts: AxiosRequestConfig = {
	httpsAgent: new https.Agent({ rejectUnauthorized: false }),
	baseURL: process.env.REACT_APP_API_URL,
};

const interceptorError = (error: AxiosError) => Promise.reject(error.response);

const instance = axios.create(opts);

instance.interceptors.response.use(response => response, interceptorError);

export default instance;
