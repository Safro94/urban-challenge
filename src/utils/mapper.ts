import { IAvailableWorker, IAvailableWorkerResponse } from '../types';

export const mapAvailableWorkers = (
	workers: IAvailableWorkerResponse[]
): IAvailableWorker[] => {
	return workers.map((worker: IAvailableWorkerResponse) => ({
		slotId: worker.slot_id,
		availableWorkersIds: worker.availableWorker_ids,
	}));
};
