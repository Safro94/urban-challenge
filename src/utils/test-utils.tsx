import { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter as Router } from 'react-router-dom';

import { theme } from '../theme';

const AllTheProviders: FC = ({ children }) => {
	return (
		<ThemeProvider theme={theme}>
			<Router>{children}</Router>
		</ThemeProvider>
	);
};

const customRender = (
	ui: ReactElement,
	options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
