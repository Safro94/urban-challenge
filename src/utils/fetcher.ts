import { AxiosResponse } from 'axios';
import {
	IAvailableWorkersResponse,
	ISlotsResponse,
	IWorkerResponse,
	RequestMethods,
} from '../types';

import axios from './axios';

interface IFetcher {
	method?: RequestMethods;
	url: string;
	data?: null | any;
	body?: null | any;
}

type Response = IWorkerResponse | ISlotsResponse | IAvailableWorkersResponse;

const fetcher = async ({
	method = RequestMethods.GET,
	url,
	body = null,
}: IFetcher): Promise<AxiosResponse<Response>> => {
	try {
		const res: AxiosResponse<Response> = await axios[method](url, body);
		return res;
	} catch (error) {
		throw error;
	}
};

export default fetcher;
