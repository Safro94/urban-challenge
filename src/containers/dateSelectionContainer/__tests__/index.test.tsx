import userEvent from '@testing-library/user-event';

import { render, screen, waitFor } from '../../../utils/test-utils';

import DateSelectionContainer from '..';

import SlotService from '../../../services/slotService';

jest.mock('../../../services/slotService', () => {
	return {
		getSlots: jest.fn(),
	};
});

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useHistory: () => ({
		push: mockHistoryPush,
	}),
}));

const mockSetSelectedSlot = jest.fn();
jest.mock('../../../hooks/application', () => {
	return {
		useApplication: () => ({
			setSelectedSlot: mockSetSelectedSlot,
		}),
	};
});

const mockHandleError = jest.fn();
jest.mock('react-error-boundary', () => ({
	...jest.requireActual('react-error-boundary'),
	useErrorHandler: () => mockHandleError,
}));

const mockAddToCart = jest.fn();
jest.mock('../../../hooks/cart', () => {
	return {
		useCart: () => ({
			addToCart: mockAddToCart,
		}),
	};
});

describe('DateSelectionContainer', () => {
	const slots = [
		{
			id: 1,
			localisedTime: '10:00',
			price: '$20',
		},
		{
			id: 2,
			localisedTime: '10:30',
			price: '$30',
		},
		{
			id: 3,
			localisedTime: '11:00',
			price: '$50',
		},
	];

	beforeEach(() => {
		(SlotService.getSlots as jest.Mock).mockResolvedValue(slots);
	});

	it('should render correctly when service returns properties', async () => {
		const { container } = render(<DateSelectionContainer />);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should call the service when the date is selected', async () => {
		render(<DateSelectionContainer />);

		const input = screen.getByRole('textbox');

		userEvent.click(input);

		userEvent.click(screen.getByText(`${new Date(Date.now()).getDate()}`));

		await waitFor(() => {
			expect(SlotService.getSlots).toHaveBeenCalledTimes(1);
			expect(SlotService.getSlots).toHaveBeenCalledWith(mockHandleError);
		});
	});

	it('should set the selected slot when the time slot button is clicked', async () => {
		render(<DateSelectionContainer />);

		const input = screen.getByRole('textbox');

		userEvent.click(input);

		userEvent.click(screen.getByText(`${new Date(Date.now()).getDate()}`));

		await waitFor(() => {
			expect(SlotService.getSlots).toHaveBeenCalledTimes(1);
			expect(SlotService.getSlots).toHaveBeenCalledWith(mockHandleError);
		});

		const timeButton = screen.getByRole('button', {
			name: `${slots[1].localisedTime} - ${slots[1].price}`,
		});

		userEvent.click(timeButton);

		expect(mockSetSelectedSlot).toHaveBeenCalledTimes(1);
		expect(mockSetSelectedSlot).toHaveBeenCalledWith(slots[1]);
	});
});
