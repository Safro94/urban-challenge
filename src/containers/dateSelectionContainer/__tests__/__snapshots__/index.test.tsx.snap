// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateSelectionContainer should render correctly when service returns properties 1`] = `
<div
  class="sc-kEjbQP kSspzS"
>
  <div
    class="sc-eCstlR evnTVf"
  >
    <div
      class="sc-hKgJUU kpfJCg"
    >
      <label
        class="sc-dlfnuX kdROzM"
      >
        Select a date
        <input
          aria-describedby="date-picker-sr-message"
          class="sc-jSgvzq udJrS"
          type="text"
          value="13/10/2021"
        />
        <div
          id="date-picker-sr-message"
          style="border: 0px; height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; position: absolute;"
        >
          <p>
            Press the down arrow key to interact with the calendar and select a date. The following keyboard shortcuts can be used to change dates.
          </p>
          <ul>
            <li>
              Enter Key: select the date in focus.
            </li>
            <li>
              Right and left arrow keys: Move backward (left) and forward (right) by one day.
            </li>
            <li>
              Up and down arrow keys: Move backward (up) and forward (down) by one week.
            </li>
            <li>
              Page up and page down keys: Switch months.
            </li>
            <li>
              Home and end keys: go to the first or last day of a week.
            </li>
            <li>
              Escape key: Return to the date input field.
            </li>
          </ul>
        </div>
      </label>
    </div>
  </div>
</div>
`;
